rate_limiter/rate_limiter.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package rate_limiter
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/time/rate"
+)
+
+// CustomRoundTripper wraps http.RoundTripper to observe metrics and rate limit if necessary
+type CustomRoundTripper struct {
+	rateLimiter *rate.Limiter
+	source      string
+	rt          http.RoundTripper
+}
+
+func (irt *CustomRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if irt.rateLimiter != nil {
+		err := irt.rateLimiter.Wait(req.Context())
+		if err != nil {
+			return nil, fmt.Errorf("transport: unable to ratelimit: %w", err)
+		}
+	}
+
+	labels := []metrics.Label{
+		{
+			Name:  "method",
+			Value: req.Method,
+		},
+		{
+			Name:  "path",
+			Value: req.URL.Path,
+		},
+		{
+			Name:  "source",
+			Value: irt.source,
+		},
+	}
+
+	defer metrics.MeasureSinceWithLabels([]string{"http", "dur"}, time.Now(), labels)
```

Comment:
This makes it so that the time we measure doesn't include the wait time for rate limiting. Should it be at the beginning to make sure that metrics can show when a plugin is getting rate limited?

Reply:
Makes sense, I will move it.